guidance/models/_model_state.py

```diff
@@ -0,0 +1,129 @@
+from __future__ import annotations
+from dataclasses import dataclass
+from typing import Optional, overload
+import base64
+import html
+
+
+@dataclass(frozen=True, slots=True)
+class Object:
+    def _html(self) -> str:
+        raise NotImplementedError
+
+    def __str__(self) -> str:
+        raise NotImplementedError
+
+
+@dataclass(frozen=True, slots=True)
+class Text(Object):
+    text: str
+    probability: Optional[float] = None
+
+    def __str__(self) -> str:
+        return self.text
+
+    def _html(self) -> str:
+        escaped_text = html.escape(self.text)
+        if self.probability is not None:
+            style = f"background-color: rgba({165*(1-self.probability) + 0}, {165*self.probability + 0}, 0, {0.15}); border-radius: 3px;"
+            return f"<span style='{style}' title='{self.probability}'>{escaped_text}</span>"
+        return escaped_text
+
+
+@dataclass(frozen=True, slots=True)
+class Image(Object):
+    id: str
+    data: bytes
+
+    def __str__(self) -> str:
+        raise NotImplementedError
+
+    def _html(self) -> str:
+        return f"""<img src="data:image/png;base64,'{base64.b64encode(self.data).decode()}'" style="max-width: 400px; vertical-align: middle; margin: 4px;">"""
+
+
+@dataclass(frozen=True, slots=True)
+class RoleOpener(Object):
+    role_name: str
+    text: str
+    indent: bool
+
+    def _html(self) -> str:
+        out = ""
+        if self.indent:
+            out += f"<div style='display: flex; border-bottom: 1px solid rgba(127, 127, 127, 0.2);  justify-content: center; align-items: center;'><div style='flex: 0 0 80px; opacity: 0.5;'>{self.role_name.lower()}</div><div style='flex-grow: 1; padding: 5px; padding-top: 10px; padding-bottom: 10px; margin-top: 0px; white-space: pre-wrap; margin-bottom: 0px;'>"
+        else:
+            out += "<span style='background-color: rgba(255, 180, 0, 0.3); border-radius: 3px;'>"
+            out += self.text
+            out += "</span>"
+        return out
+
+    def __str__(self) -> str:
+        return self.text
+
+
+@dataclass(frozen=True, slots=True)
+class RoleCloser(Object):
+    role_name: str
+    text: str
+    indent: bool
+
+    def _html(self) -> str:
+        out = ""
+        if self.indent:
+            out += "</div></div>"
+        else:
+            out += "<span style='background-color: rgba(255, 180, 0, 0.3); border-radius: 3px;'>"
+            out += self.text
+            out += "</span>"
+        return out
+
+    def __str__(self) -> str:
+        return self.text
+
+
+class ModelState:
+    __slots__ = ["objects"]
+
+    def __init__(self) -> None:
+        self.objects: list[Object] = []
```

Comment:
I imagine there are constraints on this list - for example, a `RoleCloser` _must_ be followed by a `RoleOpener`?

Reply:
Closers may come arbitrarily far in the future, but there is logic inside the model class to ensure that they get closed... Currently we just have run-time checks for balanced openers/closers inside the chat models, and I'm not sure if we can do much better... But we may at least be able to have constraints that you can't open a role if one is already open

Reply:
I was wondering if there should be a nested structure - a list of `Role` blocks, each of which could contain an arbitrary list of `Text`, `Image` etc. blocks. Or would that be overly complicated?

Reply:
I had the same thought actually. On my first attempt, that felt pretty complicated, but it might not be so bad now that this skeleton is here. I'll experiment with it a bit :)